component/src/main/java/org/ballerinalang/kafka/nativeimpl/producer/action/AbstractTransactionHandler.java

```diff
@@ -19,66 +19,74 @@
-                               Map<TopicPartition, OffsetAndMetadata> partitionToMetadataMap,
-                               String groupID) {
-        try {
-            if (isTransactionalProducer(producerProperties)) {
-                initiateTransaction(producerConnector);
-            }
-            producer.sendOffsetsToTransaction(partitionToMetadataMap, groupID);
-        } catch (IllegalStateException | KafkaException e) {
-            throw new BallerinaException("Failed to send offsets to transaction. " + e.getMessage(), e, context);
-        }
+    boolean isTransactionalProducer() {
+        return Objects.nonNull(
+                producerProperties.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG))
+                && context.isInTransaction();
     }
 
-    @Override
-    public boolean isBlocking() {
-        return false;
+    boolean isKafkaTransactionInitiated(TransactionLocalContext localTransactionInfo, String connectorKey) {
+        return Objects.nonNull(localTransactionInfo.getTransactionContext(connectorKey));
     }
 
-    private void performTransaction(LocalTransactionInfo localTransactionInfo, String connectorKey) {
-        if (!isKafkaTransactionInitiated(localTransactionInfo, connectorKey)) {
-            KafkaTransactionContext txContext = new KafkaTransactionContext(producer);
-            localTransactionInfo.registerTransactionContext(connectorKey, txContext);
-            producer.beginTransaction();
+    void initializeClassVariables() {
+        producerConnector = (BMap<String, BValue>) context.getRefArgument(0);
+        BMap producerMap = (BMap) producerConnector.get("producerHolder");
+        producerStruct = (BMap<String, BValue>) producerMap.get(new BString(NATIVE_PRODUCER));
+        producer = (KafkaProducer) producerStruct.getNativeData(NATIVE_PRODUCER);
+        producerProperties = (Properties) producerStruct.getNativeData(NATIVE_PRODUCER_CONFIG);
+    }
+
+    void commitConsumer(Map<TopicPartition, OffsetAndMetadata> partitionToMetadataMap, String groupID) {
+        if (isTransactionalProducer()) {
+            initiateTransaction();
         }
+        producer.sendOffsetsToTransaction(partitionToMetadataMap, groupID);
     }
 
-    public void initiateTransaction(BMap<String, BValue> producerConnector) {
+    void initiateTransaction() {
         String connectorKey = producerConnector.get("connectorID").stringValue();
-        LocalTransactionInfo localTransactionInfo = context.getLocalTransactionInfo();
+        TransactionLocalContext localTransactionInfo = context.getLocalTransactionInfo();
         performTransaction(localTransactionInfo, connectorKey);
     }
 
-    public boolean isTransactionalProducer(Properties properties) {
-        return Objects.nonNull(properties.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG)) && context.isInTransaction();
+    private void performTransaction(TransactionLocalContext localTransactionInfo, String connectorKey) {
```

Comment:
Shouldn't it more meaningful to use beginTransaction as the method name rather performTransaction??

Reply:
Agreed. Will do the changes accordingly.